### ISSUE TITLE
BC break in 2.3.x?

### ISSUE THREAD

**Issue by luispabon** [Ocramius/ProxyManager#464]
Looks like it requires php 7.4, which seems like a bc break that would perhaps demand a 3.0 release instead.

**Comment by Majkl578** [comment]
Changing PHP requirement is not a BC break: https://www.doctrine-project.org/2017/07/25/php-7.1-requirement-and-composer.html

**Comment by luispabon** [comment]
I think that's debatable :man_shrugging:  If I can't drop it into a working setup and have it work, then it's not backwards compatible with the version before.

**Comment by Ocramius** [comment]
Closing as invalid and duplicate of https://github.com/Ocramius/ProxyManager/issues/455

**Comment by alex2005git** [comment]
Just wanted to let you know that this is still an issue that should be addressed.
One fine day I find myself with breaking code because a library uses yours, and it broke.

@Taluu's comment is just plain wrong, `^` means all minor changes less than the major one, so ^2.0.2 will indeed select 2.4 (^2.0 is the same, it just adds the versions before 2.0.2 to the list), what he was thinking about is `~`, so `~2.0.2` would update the patch less than the minor, so 2.1 will not be selected.
You can easily see this for yourself
`^2.0.2` https://semver.mwl.be/#!?package=ocramius%2Fproxy-manager&version=%5E2.0.2&minimum-stability=stable
`~2.0.2` https://semver.mwl.be/#!?package=ocramius%2Fproxy-manager&version=~2.0.2&minimum-stability=stable

Also his comment about ignoring platform reqs doesnt take into consideration that some of us use docker images that dont have the extensions needed in their image, so we must ignore the reqs (even though we have them in the images that run php itself)

I would kindly ask you to reconsider this, if you want to be a library that follows the semver notation.

**Comment by Ocramius** [comment]
> Just wanted to let you know that this is still an issue that should be addressed.

No, it shouldn't: @Majkl578 made it clear in https://github.com/Ocramius/ProxyManager/issues/464#issuecomment-492834456.

 > One fine day I find myself with breaking code because a library uses yours, and it broke.

It broke because you were running against a dependency set not compatible with your `composer.lock`. `composer.lock` **MUST** be honored: if you don't do that, you are developing on dangerous grounds, and you should fix that.

 > @Taluu's comment is just plain wrong

@Taluu is indeed mistaken, and that happens (I also misread their comment when I locked the thread): you use `^` to target everything within a major release, and then rely on the SAT process to resolve a compatible set of dependencies and libraries matching your environment and project dependencies. This will indeed target any release within `2.x`, which is correct.

 > Also his comment about ignoring platform reqs doesnt take into consideration that some of us use docker images that dont have the extensions needed in their image, so we must ignore the reqs (even though we have them in the images that run php itself)

You **MUST** run `composer update` within an environment that reflects production. I do use docker myself, and that makes it even easier: no excuses for using a mismatching environment to generate a `composer.lock`

 > I would kindly ask you to reconsider this, if you want to be a library that follows the semver notation.

This library follows SemVer quite closely, to the point that we had to revert some BC breaks in unlikely inheritance endpoints in 2.31 (https://github.com/Ocramius/ProxyManager/pull/452).

If you want to help further, I suggest adding https://github.com/Roave/BackwardCompatibilityCheck (which I also maintain) to the build of this library: that would prevent more BC breaks like #452.

Upgrading dependencies is **not** a BC break, unless there are inherited dependency symbols that broke BC themselves (not the case here).

**Comment by alex2005git** [comment]
The whole argument seems to rely on doctrine's response about it.
But just because they say it, it doesnt mean, for me at least, that everyone swears by it.

The code broke yes, but I didn't say it was in production.  

For me, it seems simple enough.  
If I could just drop the new version in my composer file and the code would run, it's a minor bump.  
If for whatever reason it doesnt work, then it's a breaking change and should require a major version.

Sad to see this response honestly, locking threads and not listening to the people that do use your library.

**Comment by Ocramius** [comment]
> But just because they say it, it doesnt mean, for me at least, that everyone swears by it.

I am part of the folks that wrote that article: I'm not using it as a "doctrine does it, hence it is right", it is just a correct approach. Other tools cannot do this for other reasons, see for example https://github.com/zendframework/maintainers/issues/59, in which ZF can't upgrade due to LTS having to be **installable** with older PHP releases too.

 > If I could just drop the new version in my composer file and the code would run, it's a minor bump.

You'll have to upgrade dependencies, not code. This is expected, and if dependencies are locked (system dependencies), then `composer update` fails as expected, or locks you onto a release of the library that can be installed with older dependencies.

 > Sad to see this response honestly, locking threads and not listening to the people that do use your library.

This is an overall healthy approach to dependency bumps, you just happen to learn it the very hard way:

 1. `composer update` **MUST** be run in production-alike environment, no differences allowed
 2. upgrade dependencies, often
 3. the ecosystem gains from this by pushing newer versions (want the new features? need new releases!)

This is not just a selfish view: if we maintainers weren't doing this, then adoption of newer releases of anything would be much worse, and maintenance burden much higher too.

**Comment by alex2005git** [comment]
Thanks for taking the time to respond.
I respect your views, but will kindly disagree, 7.4 hasn't even been released yet.  

What happens to any future bug fixes, until 7.4 gets released in November (I would think no one has it in their production environment yet), will you maintain the version prior to the 7.4 changes?  
Will you release `2.2.3` or are we just screwed? If so then `2.3` would be like a major version for your project (which defeats the purpose of SemVer)

**Comment by Ocramius** [comment]
> What happens to any future bug fixes, until 7.4 gets released in November (I would think no one has it in their production environment yet), will you maintain the version prior to the 7.4 changes?

No, unless they cause security issues, patches for that will only go to `2.3.x`, probably even `2.4.x` (currently working on that).

 > will you maintain the version prior to the 7.4 changes?

Only security fixes.

 > Will you release 2.2.3 or are we just screwed?

I'm already working on that, see #469.

 > If so then 2.3 would be like a major version for your project (which defeats the purpose of SemVer)

Sorry, but that's you not relying on SemVer.

You should pin to `^2.2.2`, since I'm extremely careful about not introducing any BC breaks in any of the projects that I maintain, to the point that I wrote [a whole library](https://github.com/Roave/BackwardCompatibilityCheck) about just ensuring that, and I and @asgrim wrote an entire [reflection project](https://github.com/Roave/BetterReflection) over the period of two years just to get to that point.

**Comment by alex2005git** [comment]
> You should pin to ^2.2.2
Small fix, it should be `~2.2.2`, so it bumps just the patch.

I already had to do that in my composer projects that don't event include this dependency.
And we are talking about a lot of microservices that would need MR's and explanations about why it is needed.

> Sorry, but that's you not relying on SemVer.
Actually it's the other third party libraries that dont rely on SemVer, and they all should go an pin `~2.2.2` in their dependencies.

To me all this seems way too much of a hassle, won't you agree? Especially for something that hasn't even been fully released.

**Comment by Ocramius** [comment]
> Small fix, it should be ~2.2.2, so it bumps just the patch.

No, it should be `^2.2.2`, so you'll get `2.3.1` and newer once SAT decides that your environment is ready.

 > And we are talking about a lot of microservices that would need MR's and explanations about why it is needed.

It's in fact not needed: fix how you produce `composer.lock`

 > To me all this seems way too much of a hassle, won't you agree? Especially for something that hasn't even been fully released.

It is a project I care for deeply: you are free to fork and make it your own problem.

Locking here: this discussion is going nowhere, while you don't want to fix the root cause (broken build pipeline on your end).
